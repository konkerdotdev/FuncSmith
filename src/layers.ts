/* eslint-disable @typescript-eslint/naming-convention */
import * as P from '@konker.dev/effect-ts-prelude';
import { MemFsTinyFileSystem, NodeTinyFileSystem } from '@konker.dev/tiny-filesystem-fp';

import { EMPTY_FILESET, FuncSmithContextSink, FuncSmithContextSource } from './index';
import type { FileSet, FileSetItem } from './lib/fileSet';
import { fsFileSinkWriter } from './sink/FsFileSinkWriter';
import { fsFileSourceReader } from './source/FsFileSourceReader';
import memFs1 from './test/fixtures/memfs-1.json';
import {
  FuncSmithContext,
  FuncSmithContextEnv,
  FuncSmithContextFs,
  FuncSmithContextMetadata,
  FuncSmithContextReader,
  FuncSmithContextWriter,
} from './types';

// --------------------------------------------------------------------------
export const FuncSmithContextFsLive = P.Layer.succeed(
  FuncSmithContextFs,
  FuncSmithContextFs.of({
    tinyFs: NodeTinyFileSystem,
  })
);
export const FuncSmithContextFsTest = P.Layer.succeed(
  FuncSmithContextFs,
  FuncSmithContextFs.of({
    tinyFs: MemFsTinyFileSystem(memFs1, '/tmp'),
  })
);

// --------------------------------------------------------------------------
export const FuncSmithContextSourceTest = FuncSmithContextSource.of({
  sourcePath: '/tmp/foo',
});

export const FuncSmithContextSinkTest = FuncSmithContextSink.of({
  sinkPath: '/tmp/build',
});

// --------------------------------------------------------------------------
export const FuncSmithContextReaderLayer = P.Layer.effect(
  FuncSmithContextReader,
  P.Effect.map(FuncSmithContextFs, (funcSmithContextFs) =>
    FuncSmithContextReader.of({
      tinyFs: funcSmithContextFs.tinyFs,
      reader: (sourcePath: string, globPattern: string | undefined) =>
        P.pipe(
          fsFileSourceReader(sourcePath, globPattern),
          P.Effect.provideService(FuncSmithContextFs, funcSmithContextFs)
        ),
    })
  )
);
export const FuncSmithContextReaderLive = P.pipe(FuncSmithContextReaderLayer, P.Layer.provide(FuncSmithContextFsLive));
export const FuncSmithContextReaderTest = P.pipe(FuncSmithContextReaderLayer, P.Layer.provide(FuncSmithContextFsTest));

// --------------------------------------------------------------------------
export const FuncSmithContextWriterLayer = P.Layer.effect(
  FuncSmithContextWriter,
  P.Effect.map(FuncSmithContextFs, (funcSmithContextFs) =>
    FuncSmithContextWriter.of({
      tinyFs: funcSmithContextFs.tinyFs,
      writer: (sinkPath: string, fileSet: FileSet<FileSetItem>) =>
        P.pipe(fsFileSinkWriter(sinkPath, fileSet), P.Effect.provideService(FuncSmithContextFs, funcSmithContextFs)),
    })
  )
);

export const FuncSmithContextWriterLive = P.pipe(FuncSmithContextWriterLayer, P.Layer.provide(FuncSmithContextFsLive));
export const FuncSmithContextWriterTest = P.pipe(FuncSmithContextWriterLayer, P.Layer.provide(FuncSmithContextFsTest));

// --------------------------------------------------------------------------
export const FuncSmithContextEnvDefault = P.Layer.succeed(
  FuncSmithContextEnv,
  FuncSmithContextEnv.of({
    env: {},
  })
);

// --------------------------------------------------------------------------
export const FuncSmithContextMetadataDefault = P.Layer.succeed(
  FuncSmithContextMetadata,
  FuncSmithContextMetadata.of({
    metadata: {},
  })
);

export const FuncSmithContextMetadataTest = P.Layer.succeed(
  FuncSmithContextMetadata,
  FuncSmithContextMetadata.of({
    metadata: {
      testMetaDataName1: 'testMetaDataValue1',
    },
  })
);

// --------------------------------------------------------------------------
export const FuncSmithContextTest = P.Layer.succeed(
  FuncSmithContext(),
  FuncSmithContext().of({
    rootDirPath: '/tmp',
    fileSet: EMPTY_FILESET(),
  })
);
