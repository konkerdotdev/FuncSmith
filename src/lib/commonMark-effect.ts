import * as P from '@konker.dev/effect-ts-prelude';
import { toTinyError } from '@konker.dev/tiny-error-fp';
import * as M from 'commonmark';

// --------------------------------------------------------------------------
export const ERROR_TAG = 'CommonMarkError';
export type ERROR_TAG = typeof ERROR_TAG;

export const toCommonMarkError = toTinyError<ERROR_TAG>(ERROR_TAG);
export type CommonMarkError = ReturnType<typeof toCommonMarkError>;

// --------------------------------------------------------------------------
export const commonMarkParse =
  (options: M.ParserOptions = {}) =>
  (markdown: string): P.Effect.Effect<never, CommonMarkError, M.Node> =>
    P.Effect.try({ try: () => new M.Parser(options).parse(markdown), catch: toCommonMarkError });

export const commonMarkRender =
  (options: M.HtmlRenderingOptions = {}) =>
  (node: M.Node): P.Effect.Effect<never, CommonMarkError, string> =>
    P.Effect.try({ try: () => new M.HtmlRenderer(options).render(node), catch: toCommonMarkError });
