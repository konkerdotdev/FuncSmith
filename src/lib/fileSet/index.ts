import * as P from '@konker.dev/effect-ts-prelude';

import type { FrontMatter } from '../frontMatter';

export enum FileSetItemType {
  File = 'File',
}

export type FileSetItemFile<T = ArrayBuffer | string> = {
  _tag: FileSetItemType.File;
  path: string;
  baseDir: string;
  relPath: string;
  relDir: string;
  fileName: string;
  fileBase: string;
  fileExt: string;
  contents: T;
};

export type FileSetItem<T = ArrayBuffer | string> = FileSetItemFile<T>;

export type FileSet<T extends FileSetItem<C>, C = ArrayBuffer | string> = Array<T>;

// --------------------------------------------------------------------------
export const HtmlString = P.pipe(P.Schema.string, P.Schema.brand(Symbol.for('HtmlString')));
export type HtmlString = P.Schema.Schema.To<typeof HtmlString>;
export type Html<T extends FileSetItem> = FrontMatter<Omit<T, 'contents'>> & {
  readonly contents: HtmlString;
};
