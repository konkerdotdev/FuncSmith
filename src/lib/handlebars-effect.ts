import * as P from '@konker.dev/effect-ts-prelude';
import { toTinyError } from '@konker.dev/tiny-error-fp';
import H from 'handlebars';

// --------------------------------------------------------------------------
export const ERROR_TAG = 'HandlebarsError';
export type ERROR_TAG = typeof ERROR_TAG;

export const toHandlebarsError = toTinyError<ERROR_TAG>(ERROR_TAG);
export type HandlebarsError = ReturnType<typeof toHandlebarsError>;

// --------------------------------------------------------------------------
export function handlebarsCompile(
  templateStr: string,
  helpers: Record<string, H.HelperDelegate> = {}
): P.Effect.Effect<never, HandlebarsError, H.TemplateDelegate> {
  return P.Effect.try({
    try: () => {
      // eslint-disable-next-line fp/no-unused-expression
      Object.keys(helpers).forEach((name) => H.registerHelper(name, helpers[name]!));
      return H.compile(templateStr);
    },
    catch: toHandlebarsError,
  });
}

export const handlebarsRender =
  (context: unknown) =>
  (template: H.TemplateDelegate): P.Effect.Effect<never, HandlebarsError, string> =>
    P.Effect.try({
      try: () => template(context),
      catch: toHandlebarsError,
    });

export function handlebarsRender2(
  context: unknown,
  template: H.TemplateDelegate
): P.Effect.Effect<never, HandlebarsError, string> {
  return handlebarsRender(context)(template);
}

export const handlebars =
  (templateStr: string, helpers: Record<string, H.HelperDelegate> = {}) =>
  (context: unknown): P.Effect.Effect<never, HandlebarsError, string> =>
    P.pipe(handlebarsCompile(templateStr, helpers), P.Effect.flatMap(handlebarsRender(context)));
