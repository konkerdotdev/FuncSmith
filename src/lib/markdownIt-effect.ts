import * as P from '@konker.dev/effect-ts-prelude';
import { toTinyError } from '@konker.dev/tiny-error-fp';
import H from 'highlight.js';
import markdownIt from 'markdown-it';

import type { MarkdownOptions } from '../plugins/Markdown/types';

// --------------------------------------------------------------------------
export const ERROR_TAG = 'MarkdownItError';
export type ERROR_TAG = typeof ERROR_TAG;

export const toMarkdownItError = toTinyError<ERROR_TAG>(ERROR_TAG);
export type MarkdownItError = ReturnType<typeof toMarkdownItError>;

// --------------------------------------------------------------------------
export function adaptOptions(_options: MarkdownOptions | undefined): markdownIt.Options {
  return {
    highlight: function (str, lang) {
      if (lang && H.getLanguage(lang)) {
        try {
          return H.highlight(lang, str, true).value;
        } catch (__) {}
      }

      return ''; // use external default escaping
    },
  };
}

// --------------------------------------------------------------------------
export const markdownItRender =
  (options?: MarkdownOptions) =>
  (markdown: string): P.Effect.Effect<string, MarkdownItError> =>
    P.Effect.tryPromise({
      try: async () => markdownIt(adaptOptions(options)).render(markdown),
      catch: toMarkdownItError,
    });
