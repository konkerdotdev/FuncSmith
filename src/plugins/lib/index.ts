import * as P from '@konker.dev/effect-ts-prelude';

import type { FileSet, FileSetItem } from '../../lib/fileSet';
import type { FileSetInjectionMappingCtor, FileSetMapping, FileSetMappingCtor } from '../../types';

/**
 * A type when the plugin wants to accept either a full options record, or a string convenience version.
 * E.g.
 * ```
 *     F.collections({
 *       posts: {
 *         globPattern: 'posts/*.html',
 *         sortBy: 'date',
 *       },
 *     }),
 * ```
 * vs
 * ```
 *     F.collections({
 *       posts: 'posts/*.html',
 *     }),
 * ```
 */
export type Convenience<T> = string | Partial<T>;

/**
 * IF - input fileSetItem
 * OF - output fileSetItem
 * MF - mapping output fileSetItem
 * RM - mapping deps
 * RN - next mapping deps
 * OP - options
 */
export const wrapMapping =
  <IF extends FileSetItem, MF extends FileSetItem, RM, OP = undefined>(
    mappingCtor: FileSetMappingCtor<IF, MF, RM, OP>
  ) =>
  (options?: OP) =>
  <OF extends FileSetItem, RN>(next: FileSetMapping<MF, OF, RN>) =>
  (fileSet: FileSet<IF>) =>
    P.pipe(P.Effect.succeed(fileSet), P.Effect.flatMap(mappingCtor(options)), P.Effect.flatMap(next));

/**
 * IF - input fileSetItem
 * OF - output fileSetItem
 * MF - mapping output fileSetItem
 * RM - mapping deps
 * RN - next mapping deps
 * OP - options
 */
export const wrapInjection =
  <IF extends FileSetItem, OF extends FileSetItem, RN, RM, OP = undefined>(
    postMappingCtor: FileSetInjectionMappingCtor<OF, RN, RM, OP>
  ) =>
  (options: OP) =>
  (next: FileSetMapping<IF, OF, RN>): FileSetMapping<IF, OF, RM> =>
  (fileSet: FileSet<IF>) =>
    P.pipe(next(fileSet), postMappingCtor(options));
