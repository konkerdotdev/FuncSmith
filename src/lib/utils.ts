import * as P from '@konker.dev/effect-ts-prelude';
import { toTinyError } from '@konker.dev/tiny-error-fp';
import { stringToUint8Array } from '@konker.dev/tiny-filesystem-fp/dist/lib/array';
import * as crypto from 'crypto';

// --------------------------------------------------------------------------
export const ERROR_TAG = 'GeneralError';
export type ERROR_TAG = typeof ERROR_TAG;

export const toGeneralError = toTinyError<ERROR_TAG>(ERROR_TAG);
export type GeneralError = ReturnType<typeof toGeneralError>;

// --------------------------------------------------------------------------
export function hashHex(s: string): P.Effect.Effect<string, GeneralError> {
  return P.pipe(
    P.Effect.tryPromise({
      try: async () => crypto.subtle.digest('SHA-1', stringToUint8Array(s)),
      catch: toGeneralError,
    }),
    P.Effect.map((buffer) => [...new Uint8Array(buffer)].map((x) => x.toString(16).padStart(2, '0')).join(''))
  );
}

// --------------------------------------------------------------------------
export function arrayMapInPlace<A, B>(
  aa: ReadonlyArray<A>,
  fn: (a: A, i: number, arr: ReadonlyArray<A | B>) => B
): ReadonlyArray<A | B> {
  return aa.reduce(
    (acc: ReadonlyArray<A | B>, val: A, i: number, arr: ReadonlyArray<A>) =>
      acc
        .slice(0, i)
        .concat(fn(val, i, acc))
        .concat(arr.slice(i + 1)),
    aa as ReadonlyArray<A>
  );
}

// --------------------------------------------------------------------------S
export { arrayBufferToString, stringToUint8Array } from '@konker.dev/tiny-filesystem-fp/dist/lib/array';
