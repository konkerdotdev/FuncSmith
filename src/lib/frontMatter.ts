// --------------------------------------------------------------------------
import * as P from '@konker.dev/effect-ts-prelude';
import grayMatter from 'gray-matter';

import type { FuncSmithError } from '../error';
import { toFuncSmithError } from '../error';
import type { FrontMatter } from '../plugins/FrontMatter/types';
import type { FileSetItem, FileSetItemFile } from './fileSet';
import { fileSetItemContentsToString } from './fileSet/fileSetItem';

export function isFrontMatter<T extends FileSetItem>(fileSetItem: T): fileSetItem is FrontMatter<T> {
  return 'frontMatter' in fileSetItem && typeof fileSetItem.frontMatter === 'object';
}

export const extractFrontMatter =
  (extraFrontMatter: Record<string, unknown> = {}) =>
  <T extends FileSetItemFile>(fileSetItem: T): P.Effect.Effect<FrontMatter<T>, FuncSmithError> =>
    P.Effect.try({
      try: () => {
        const { content, data } = grayMatter(fileSetItemContentsToString(fileSetItem.contents));
        return {
          ...fileSetItem,
          ...data,
          ...extraFrontMatter,
          frontMatter: {
            ...data,
            ...extraFrontMatter,
          },
          contents: content,
        };
      },
      catch: toFuncSmithError,
    });
