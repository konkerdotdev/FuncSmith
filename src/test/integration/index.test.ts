/* eslint-disable fp/no-unused-expression,fp/no-nil,fp/no-mutation,fp/no-let */
import * as P from '@konker.dev/effect-ts-prelude';
import { fs } from 'memfs';
import { toTreeSync } from 'memfs/lib/print';

import type { FileSetMapping, FuncSmithPlugin } from '../../index';
import * as F from '../../index';
import { FsDepReaderTest, FsDepWriterTest } from '../../layers';
import type { FileSetItem } from '../../lib/fileSet';

describe('funcsmith', () => {
  let env: any;
  let metadata: any;

  function useImpl<IF extends FileSetItem, OF extends FileSetItem, RN, RM>(plugin1: FuncSmithPlugin<IF, OF, RN, RM>) {
    return function (plugin2: FuncSmithPlugin<IF, OF, RN, RM>) {
      return P.flow(plugin1, plugin2);
    };
  }
  // function use<IF extends FileSetItem, OF extends FileSetItem, R>(mapping: FileSetMapping<IF, OF, R>) {
  //   return { use: mapping };
  // }

  const x = F.root(__dirname)(F.source('/tmp/foo')(F.sink('/tmp/www')(P.Effect.succeed)));
  const xx = P.flow(F.root(__dirname), F.source('/tmp/foo'), F.sink('/tmp/www'));
  const xxx = xx(P.Effect.succeed);

  test('integration', async () => {
    const pluginStack = P.flow(
      F.root(__dirname),
      F.source('/tmp/foo'),
      F.sink('/tmp/www'),
      F.cleaner(),
      F.metadata({
        siteName: 'My Static Site & Blog',
        siteUrl: 'https://example.com/',
        description: "It's about saying »Hello« to the world.",
        generatorName: 'funcsmith',
        generatorUrl: 'https://konker.dev/funcsmith/',
      }),
      F.reader(),
      F.frontMatter(),
      F.markdown(),
      F.writer()
    );

    const actual = await P.Effect.runPromise(
      P.pipe(F.FuncSmith2(pluginStack), P.Effect.provide(FsDepReaderTest), P.Effect.provide(FsDepWriterTest))
    );

    expect(actual).toMatchSnapshot('integration-test-1');
    expect(toTreeSync(fs)).toMatchSnapshot('integration-test-1-fs');
    expect(env).toMatchSnapshot('integration-test-env-1-env');
    expect(metadata).toMatchSnapshot('integration-test-env-1-metadata');
  });

  test('integration', async () => {
    const pluginStack = P.pipe(
      P.Effect.succeed,
      F.writer(),
      F.debug((fsDepEnv, fsDepMetadata, fileSet) => {
        console.log('DEBUG', fileSet.length);
        env = fsDepEnv;
        metadata = fsDepMetadata;
      }),
      F.filter(),
      F.layouts({
        templateEngine: 'handlebars',
        layoutsPath: '/tmp/layouts',
        partialsPath: '/tmp/partials',
        helpers: {
          formattedDate: function (date) {
            return new Date(date).toLocaleDateString();
          },
          toJSON: function (x) {
            // eslint-disable-next-line fp/no-nil
            return JSON.stringify(x, null, 2);
          },
        },
        globPattern: '**/*.html',
      }),
      F.nav(),
      F.collections({
        posts: {
          globPattern: 'posts/**/*.html',
          sortBy: 'date',
        },
      }),
      F.permalinks({
        match: ['**/posts/**/*.html'],
        directoryIndex: 'index.html',
        trailingSlash: false,
      }),
      F.rename([[/\.md$/, '.html']]),
      F.markdown(),
      F.drafts(),
      F.identity(),
      F.frontMatter(),
      F.reader(),
      F.metadata({
        siteName: 'My Static Site & Blog',
        siteUrl: 'https://example.com/',
        description: "It's about saying »Hello« to the world.",
        generatorName: 'funcsmith',
        generatorUrl: 'https://konker.dev/funcsmith/',
      }),
      F.env({
        TEST_SOME_ENV_VAR: 'TEST_SOME_ENV_VAR_VALUE',
      }),
      F.cleaner(),
      F.sink('/tmp/www'),
      F.source('/tmp/foo'),
      F.root(__dirname)
    );

    const actual = await P.Effect.runPromise(
      P.pipe(F.FuncSmith(pluginStack), P.Effect.provide(FsDepReaderTest), P.Effect.provide(FsDepWriterTest))
    );

    expect(actual).toMatchSnapshot('integration-test-1');
    expect(toTreeSync(fs)).toMatchSnapshot('integration-test-1-fs');
    expect(env).toMatchSnapshot('integration-test-env-1-env');
    expect(metadata).toMatchSnapshot('integration-test-env-1-metadata');
  });
});
