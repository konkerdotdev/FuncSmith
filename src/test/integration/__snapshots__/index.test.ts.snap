// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`funcsmith integration: integration-test-1 1`] = `
[
  {
    "_id": "e57ec8617d467edfff00943eac189e0ec6eb1875",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "contents": Uint8Array [
      65,
    ],
    "fileBase": "a",
    "fileExt": ".txt",
    "fileName": "a.txt",
    "link": "/a.txt",
    "path": "/tmp/foo/a.txt",
    "relDir": ".",
    "relPath": "a.txt",
  },
  {
    "_id": "39c8584020d82dc00bda6e549a6c55284dd2f5f3",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "contents": Uint8Array [
      66,
    ],
    "fileBase": "b",
    "fileExt": ".txt",
    "fileName": "b.txt",
    "link": "/b.txt",
    "path": "/tmp/foo/b.txt",
    "relDir": ".",
    "relPath": "b.txt",
  },
  {
    "_id": "c7712f15959be6b0c2ad22b5b71d2e367d6ff138",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "contents": Uint8Array [
      98,
      97,
      109,
      44,
      98,
      97,
      122,
      10,
      116,
      114,
      117,
      101,
      44,
      102,
      97,
      108,
      115,
      101,
      10,
    ],
    "fileBase": "c",
    "fileExt": ".csv",
    "fileName": "c.csv",
    "link": "/c.csv",
    "path": "/tmp/foo/c.csv",
    "relDir": ".",
    "relPath": "c.csv",
  },
  {
    "_id": "c367b94579a7a9ac870a1b33a35075533487d8d6",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "contents": Uint8Array [
      123,
      34,
      98,
      97,
      109,
      34,
      58,
      32,
      116,
      114,
      117,
      101,
      44,
      32,
      34,
      98,
      97,
      122,
      34,
      58,
      32,
      32,
      102,
      97,
      108,
      115,
      101,
      32,
      125,
    ],
    "fileBase": "d",
    "fileExt": ".json",
    "fileName": "d.json",
    "link": "/d.json",
    "path": "/tmp/foo/d.json",
    "relDir": ".",
    "relPath": "d.json",
  },
  {
    "_id": "a5e1be9ba6cb5bf18597ee658159266a000426e6",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "contents": Uint8Array [
      73,
      71,
      78,
      79,
      82,
      69,
      32,
      49,
    ],
    "fileBase": "ignore1",
    "fileExt": ".txt",
    "fileName": "ignore1.txt",
    "link": "/ignore1.txt",
    "path": "/tmp/foo/ignore1.txt",
    "relDir": ".",
    "relPath": "ignore1.txt",
  },
  {
    "_id": "460eab01f1a4de403aced612ab4230ee610d6c09",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "contents": Uint8Array [
      69,
    ],
    "fileBase": "e",
    "fileExt": ".txt",
    "fileName": "e.txt",
    "link": "/bar/e.txt",
    "path": "/tmp/foo/bar/e.txt",
    "relDir": "bar",
    "relPath": "bar/e.txt",
  },
  {
    "_id": "f8cba7f6ca545e7d307a508d628ef61dc9ac8072",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "contents": Uint8Array [
      70,
    ],
    "fileBase": "f",
    "fileExt": ".log",
    "fileName": "f.log",
    "link": "/bar/f.log",
    "path": "/tmp/foo/bar/f.log",
    "relDir": "bar",
    "relPath": "bar/f.log",
  },
  {
    "_id": "5da00cfdc5bee8aa488d13af769799a7b4043766",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "collection": {
      "len": 3,
      "next": {
        "_id": "af1250e6e96150b7d36c7b0448d0a11275cddb93",
        "_tag": "File",
        "baseDir": "/tmp/foo",
        "contents": "<h1>P1</h1>
<p>p1 content</p>
",
        "date": 2024-01-01T00:00:00.000Z,
        "draft": false,
        "fileBase": "index",
        "fileExt": ".html",
        "fileName": "index.html",
        "frontMatter": {
          "date": 2024-01-01T00:00:00.000Z,
          "draft": false,
          "layout": "post.hbs",
          "title": "P1",
        },
        "layout": "post.hbs",
        "link": "/posts/p1",
        "path": "/tmp/foo/posts/p1/index.html",
        "relDir": "posts/p1",
        "relPath": "posts/p1/index.html",
        "title": "P1",
      },
    },
    "contents": "LAYOUT Home",
    "date": 2023-12-31T00:00:00.000Z,
    "fileBase": "index",
    "fileExt": ".html",
    "fileName": "index.html",
    "frontMatter": {
      "date": 2023-12-31T00:00:00.000Z,
      "layout": "layout.hbs",
      "title": "Home",
    },
    "layout": "layout.hbs",
    "link": "/posts",
    "path": "/tmp/foo/posts/index.html",
    "relDir": "posts",
    "relPath": "posts/index.html",
    "title": "Home",
  },
  {
    "_id": "af1250e6e96150b7d36c7b0448d0a11275cddb93",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "collection": {
      "len": 3,
      "next": {
        "_id": "56b92724a5ed41ac6817e76408915c758d1ce978",
        "_tag": "File",
        "baseDir": "/tmp/foo",
        "contents": "<h1>P3</h1>
<p>p3 content</p>
",
        "date": 2024-01-03T00:00:00.000Z,
        "draft": false,
        "fileBase": "index",
        "fileExt": ".html",
        "fileName": "index.html",
        "frontMatter": {
          "date": 2024-01-03T00:00:00.000Z,
          "draft": false,
          "layout": "post.hbs",
          "title": "P3",
        },
        "layout": "post.hbs",
        "link": "/posts/p3",
        "path": "/tmp/foo/posts/p3/index.html",
        "relDir": "posts/p3",
        "relPath": "posts/p3/index.html",
        "title": "P3",
      },
      "previous": {
        "_id": "5da00cfdc5bee8aa488d13af769799a7b4043766",
        "_tag": "File",
        "baseDir": "/tmp/foo",
        "contents": "<h1>Home</h1>
<p>home content</p>
",
        "date": 2023-12-31T00:00:00.000Z,
        "fileBase": "index",
        "fileExt": ".html",
        "fileName": "index.html",
        "frontMatter": {
          "date": 2023-12-31T00:00:00.000Z,
          "layout": "layout.hbs",
          "title": "Home",
        },
        "layout": "layout.hbs",
        "link": "/posts",
        "path": "/tmp/foo/posts/index.html",
        "relDir": "posts",
        "relPath": "posts/index.html",
        "title": "Home",
      },
    },
    "contents": "POST: P1",
    "date": 2024-01-01T00:00:00.000Z,
    "draft": false,
    "fileBase": "index",
    "fileExt": ".html",
    "fileName": "index.html",
    "frontMatter": {
      "date": 2024-01-01T00:00:00.000Z,
      "draft": false,
      "layout": "post.hbs",
      "title": "P1",
    },
    "layout": "post.hbs",
    "link": "/posts/p1",
    "path": "/tmp/foo/posts/p1/index.html",
    "relDir": "posts/p1",
    "relPath": "posts/p1/index.html",
    "title": "P1",
  },
  {
    "_id": "56b92724a5ed41ac6817e76408915c758d1ce978",
    "_tag": "File",
    "baseDir": "/tmp/foo",
    "collection": {
      "len": 3,
      "previous": {
        "_id": "af1250e6e96150b7d36c7b0448d0a11275cddb93",
        "_tag": "File",
        "baseDir": "/tmp/foo",
        "contents": "<h1>P1</h1>
<p>p1 content</p>
",
        "date": 2024-01-01T00:00:00.000Z,
        "draft": false,
        "fileBase": "index",
        "fileExt": ".html",
        "fileName": "index.html",
        "frontMatter": {
          "date": 2024-01-01T00:00:00.000Z,
          "draft": false,
          "layout": "post.hbs",
          "title": "P1",
        },
        "layout": "post.hbs",
        "link": "/posts/p1",
        "path": "/tmp/foo/posts/p1/index.html",
        "relDir": "posts/p1",
        "relPath": "posts/p1/index.html",
        "title": "P1",
      },
    },
    "contents": "POST: P3",
    "date": 2024-01-03T00:00:00.000Z,
    "draft": false,
    "fileBase": "index",
    "fileExt": ".html",
    "fileName": "index.html",
    "frontMatter": {
      "date": 2024-01-03T00:00:00.000Z,
      "draft": false,
      "layout": "post.hbs",
      "title": "P3",
    },
    "layout": "post.hbs",
    "link": "/posts/p3",
    "path": "/tmp/foo/posts/p3/index.html",
    "relDir": "posts/p3",
    "relPath": "posts/p3/index.html",
    "title": "P3",
  },
]
`;

exports[`funcsmith integration: integration-test-fs-1 1`] = `
"/
└─ tmp/
   ├─ foo/
   │  ├─ a.txt
   │  ├─ b.txt
   │  ├─ bar/
   │  │  ├─ e.txt
   │  │  └─ f.log
   │  ├─ c.csv
   │  ├─ d.json
   │  ├─ ignore1.txt
   │  └─ posts/
   │     ├─ index.md
   │     ├─ p1.md
   │     ├─ p2.md
   │     └─ p3.md
   ├─ layouts/
   │  ├─ bar.hbs
   │  ├─ foo.hbs
   │  ├─ layout.hbs
   │  └─ post.hbs
   ├─ partials/
   │  └─ footer.hbs
   └─ www/
      ├─ a.txt
      ├─ b.txt
      ├─ bar/
      │  ├─ e.txt
      │  └─ f.log
      ├─ c.csv
      ├─ d.json
      ├─ ignore1.txt
      └─ posts/
         ├─ index.html
         ├─ p1/
         │  └─ index.html
         └─ p3/
            └─ index.html"
`;
